iib/workers/tasks/utils.py

```diff
@@ -280,3 +280,26 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
     if arg_value is None and len(args) > arg_index:
         arg_value = args[arg_index]
     return arg_value
+
+
+def get_free_port(start_port, end_port):
```

Comment:
Are you sure on this approach of parsing netstat output to find the free port?  It doesn't seem designed for parsing.

In my experience, the usual approach for this kind of thing is just to actually open the ports.  e.g. if you tried to open port `start_port` and it succeeded, you know that port was available.  You can then close it and return that number, or increment and try the next number if it wasn't available.  Just use SO_REUSEADDR on the socket to ensure it can be reused with no timeout.

That approach is racy, but equally so as the current approach (a port which appeared unused when you ran netstat could become used immediately afterward).  Not sure exactly what are your requirements around having to use a specific range of ports in the first place, maybe there is an easy way to redesign it so there's no race conditions, or maybe not.  Retries are probably enough to deal with the race condition.

Reply:
I tried using socket.connect and socket.bind, but I couldn't get it to reliably recognize that the port is in use. I think the current solution is more reliable, even though the parsing can be a pain.

Regarding the race condition, I think that the probability of an issue is low enough that it can be left as-is. During prototyping I tried to gauge the port availability directly from the command's error message (which would solve the race condition), but had trouble catching output and making sure that the process is not left hanging. Parent method has retries, so that should minimize the probability of an issue further.

Reply:
Have you tried something like this?
https://gist.github.com/gabrielfalcao/20e567e188f588b65ba2#file-get-free-tcp-port-py-L3-L8

To workaround the race condition, you can use a loop around this function call and running the `opm` command until success or an error that doesn't relate to the port not being available.

Reply:
I have rewritten it to simply retry the service in case there's a port collision. It eliminates the race condition and also seems like the most reliable method to detect port collision.